### PR TITLE
Change the way packing Android external extensions.

### PR DESCRIPTION
**Pull request by leonhsl**
Eliminate extensions-config.json.

BUG=XWALK-5756

### CODE REVIEW

**Comment by leonhsl** [comment]
Hi, @r0b5t4  PTAL, Thanks a lot~

This PR depends on https://github.com/crosswalk-project/crosswalk/pull/3422 , which will let Crosswalk iterate assets/xwalk-extensions folder instead of parsing extensions-config.json.

File: android/lib/AndroidPlatform.js [hunk 1]
```diff
@@ -1258,28 +1255,30 @@ function() {
             return;
         }
 
+        var extRootPath = Path.join(this.platformPath, "assets", "xwalk-extensions");
+        var jsjsonDstPath = Path.join(extRootPath, extName);
```
Comment (r0b5t4) [inline android/lib/AndroidPlatform.js]: Can we just call it "jsonDstPath"?

**Comment by leonhsl** [comment]
Thanks a lot for review~ Addressed comments and pushed again, Thanks~

**Comment by r0b5t4** [comment]
Many thanks @leonhsl. I think this is ok to merge now -- but one question.

What crosswalk versions does this work with? We need to coordinate app-tools and crosswalk releases.

**Comment by leonhsl** [comment]
Thanks for acknowledgement! Yeah this PR need to coordinate with the Crosswalk PR https://github.com/crosswalk-project/crosswalk/pull/3422 which I suppose will be merged soon and will be released with Crosswalk 18.

**Comment by r0b5t4** [comment]
Ok. I will merge this PR a bit later then. We will probably make an app-tools release before. But you can consider it finished. LGTM.

**Comment by leonhsl** [comment]
Hi Robert, Thanks for kindly consideration and a reminder that the Crosswalk PR https://github.com/crosswalk-project/crosswalk/pull/3422 has been merged.

**Comment by halton** [comment]
@r0b5t4 can this PR merge?

**Comment by r0b5t4** [comment]
cc @baleboy 

@halton : We need to make a plan. The problem is that the matching crosswalk support XWALK-5746 is for crosswalk 18. If we merge now, and make an app-tools release, then packaging extensions for crosswalk versions lower than 18 will not work. App-tools needs to support stable and beta as well.

**Comment by baleboy** [comment]
I suggest that we first make an app tools release without this change, and then merge it in master

**Comment by r0b5t4** [comment]
Can we merge once 16 is in stable? Crosswalk will branch before that, so the app-tools stable 0.9 branch will support releases up to 17.

**Comment by baleboy** [comment]
sounds good

**Comment by Honry** [comment]
@r0b5t4, 16 is stable now, can we merge this PR? We desire this fix for building IAP extension app with app-tools.

**Comment by r0b5t4** [comment]
Yes, thanks.

On 15 January 2016 at 09:58, Wanming Lin <notifications@github.com> wrote:

> @r0b5t4 <https://github.com/r0b5t4>, 16 is stable now, can we merge this
> PR? We desire this fix for building IAP extension app with app-tools.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/crosswalk-project/crosswalk-app-tools/pull/85#issuecomment-171905592>
> .
>



-- 
Intel GmbH - Registergericht Muenchen HRB 47456 - Dornacher Strasse 1,
85622 Feldkirchen, Germany

**Comment by r0b5t4** [comment]
Ok.
In AndroidPlantform.js around line 320, we need to check for the correct crosswalk version.

**Comment by baleboy** [comment]
Looks like the tool fails for versions < 18. This seems a bit too restrictive.

**Comment by leonhsl** [comment]
Yeah, this PR will package extensions into apk with a new folder structure which is supported from Crosswalk 18.
Do we need to continue supporting the old folder structure before Crosswalk 18 anyway?
If so, we need to change the code to make a choice to package with the NEW or OLD folder structure by checking xwalk version during tool running. 
Thanks~

**Comment by baleboy** [comment]
we can say that extensions are only supported starting from Crosswalk 18, but right now any packaging (even without extensions) will fail. Is it possible to require Crosswalk 18 only if extensions are in use?

**Comment by leonhsl** [comment]
Yeah I agree with this logic. I think it's possible and I'd like to investigate around and update this PR. @r0b5t4 WDYT?

**Comment by r0b5t4** [comment]
@baleboy I have created a 0.9 branch. This one should support crosswalk 16 and 17, and we can make more npm releases from it. Intrusive changes needed for Crosswalk 18 can land in master. Hope this is ok for everyone.

**Comment by baleboy** [comment]
ok for me

**Comment by leonhsl** [comment]
OK.
I've updated this PR for xwalk version check, will prompt error message while it's less than 18. 
But seems the buildbot failed:
> *** ERROR: No available Crosswalk versions found
I'm supposing it is because of  appveyor environment configuration, am I right? Thanks~

**Comment by Honry** [comment]
@leonhsl, I think I can help identify the issue.

**Comment by leonhsl** [comment]
It's very helpful for me! Thanks a lot~ @Honry

**Comment by Honry** [comment]
Now Crosswalk <18 does not support with this fix. The main.js targets to check app create and build with Crosswalk stable, which also need update to align with this change.

**Comment by leonhsl** [comment]
I think when the bellowing error happened, app tools running does NOT reach my code change in AndroidPlatform.js yet. 
Does the message mean that no any Crosswalk available to download in `download01`??
> *** ERROR: No available Crosswalk versions found

**Comment by leonhsl** [comment]
@r0b5t4  With the current PR checking xwalk version must >18,  android/test/main.js always fails as it is expecting success for `"stable"` xwalk. However, there has no `"stable"` xwalk 18 now. So maybe we have 2 choices here:
1, Change android/test/main.js to test with `"canary"` xwalk, then the test will succeed with `"canary"` xwalk 18.
Or
2, Hold merging this PR until  `"stable"` xwalk 18 got released
WDYT? Thanks~

**Comment by r0b5t4** [comment]
@leonhsl I think the test should be changed. Thanks.

**Comment by leonhsl** [comment]
Updated PR to let android/test/main.js use `'canary'` crosswalk instead of default `'stable'` crosswalk.
@r0b5t4  PTAL, Thanks~

**Comment by r0b5t4** [comment]
Many thanks @leonhsl

**Pull request merged**